SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc

```diff
@@ -11,27 +11,300 @@
 				      uint32_t digitizationType, CLHEP::HepRandomEngine* engine)
 {
-  runCaliceLikeDigitizer(digiColl,simData,theGeom,validIds,engine);
+  if(algo_ == 0)
+    runEmptyDigitizer(digiColl,simData,theGeom,validIds,engine);
+  else if (algo_ == 1)
+    runCaliceLikeDigitizer(digiColl,simData,theGeom,validIds,engine);
+  else if (algo_ == 2)
+    runRealisticDigitizer(digiColl,simData,theGeom,validIds,engine);
 }
 
+void HGCHEbackDigitizer::runEmptyDigitizer(std::unique_ptr<HGCalDigiCollection> &digiColl,HGCSimHitDataAccumulator &simData,
+					   const CaloSubdetectorGeometry* theGeom, const std::unordered_set<DetId>& validIds,
+					   CLHEP::HepRandomEngine* engine)
+{
+  HGCSimHitData chargeColl, toa;
+  // this represents a cell with no signal charge
+  HGCCellInfo zeroData;
+  zeroData.hit_info[0].fill(0.f); //accumulated energy
+  zeroData.hit_info[1].fill(0.f); //time-of-flight
+
+  for( const auto& id : validIds ) {
+
+    chargeColl.fill(0.f);
+    toa.fill(0.f);
+    HGCSimHitDataAccumulator::iterator it = simData.find(id);
+    HGCCellInfo& cell = ( simData.end() == it ? zeroData : it->second );
+    addCellMetadata(cell,theGeom,id);
+
+    for(size_t i=0; i<cell.hit_info[0].size(); ++i)
+    {
+      //convert total energy keV->MIP, since converted to keV in accumulator
+      const float totalIniMIPs( cell.hit_info[0][i]*keV2MIP_ );
+
+      //store
+      chargeColl[i] = totalIniMIPs;
+    }
+
+    //init a new data frame and run shaper
+    HGCalDataFrame newDataFrame( id );
+    this->myFEelectronics_->runShaper( newDataFrame, chargeColl, toa, 1, engine );
+
+    //prepare the output
+    this->updateOutput(digiColl,newDataFrame);
+  }
+}
+
+void HGCHEbackDigitizer::runRealisticDigitizer(std::unique_ptr<HGCalDigiCollection> &digiColl,HGCSimHitDataAccumulator &simData,
+  const CaloSubdetectorGeometry* theGeom, const std::unordered_set<DetId>& validIds,
+  CLHEP::HepRandomEngine* engine)
+  {
+    //switch to true if you want to print some details
+    constexpr bool debug(false);
+
+    HGCSimHitData chargeColl, toa;
+    // this represents a cell with no signal charge
+    HGCCellInfo zeroData;
+    zeroData.hit_info[0].fill(0.f); //accumulated energy
+    zeroData.hit_info[1].fill(0.f); //time-of-flight
+
+    // needed to compute the radiation and geometry scale factors
+    scal_.setGeometry(theGeom);
+
+    for( const auto& id : validIds ) {
+
+      chargeColl.fill(0.f);
+      toa.fill(0.f);
+      HGCSimHitDataAccumulator::iterator it = simData.find(id);
+      HGCCellInfo& cell = ( simData.end() == it ? zeroData : it->second );
+      addCellMetadata(cell,theGeom,id);
+
+      for(size_t i=0; i<cell.hit_info[0].size(); ++i)
+      {
+        //convert total energy keV->MIP, since converted to keV in accumulator
+        float totalIniMIPs( cell.hit_info[0][i]*keV2MIP_ );
+
+        //take into account the different size of the tiles
+        float scaledPePerMip = nPEperMIP_;
+        if(scaleByArea_)
```

Comment:
try to avoid doing `computeRadius()` twice when both this option and `scaleByDose_` are enabled:
```
double radius = 0.;
if(scaleByArea_ or scaleByDose_) radius = scal_.computeRadius(id);
```
then pass the value to `scaleByArea()` and `scaleByDose()`